### DISCUSSION TITLE
how to write a rule that make the result to be a function of input？

### DISCUSSION THREAD

**Discussion by big91987** [egraphs-good/egg#209] [Q&A]
Hi Everyone, 
Recently I was working using the egg for quantum gate compiling problem and encountered this scene: 
I need the rfs of rule  after '=>' be a function of the inputs, in the julia version of egg It's work，like this：

``` julia code ````  (https://github.com/overshiki/SymbolicCircuit.jl/blob/main/src/rule.jl)
commute_rule = @rule a b a::Gate * b::Gate => :($(b) * $(a)) where is_commute(a, b)
cancel_rule = @rule a b a::Gate * b::Gate => One() where is_cancel(a, b)
expand_rule = @rule a a::Gate => expand(a) where is_expand(a)
merge_rule = @rule a b a::Gate * b::Gate => merge(a, b) where is_merge(a, b)

but I can not find A corresponding implementation in this rust version egg，I read the code in tests and found the Analysis Trait maybe work but not very sure （It looks more like a trait for cost calculation method），Is there any solution? Thank you ：）

**Comment by big91987** [comment]
the julia version egg is Metatheory.jl, we use this julia impl and encounter some performance issue and we want to use the rust  version to implement this 。。

**Reply by mwillsey** [reply]
You might be looking for a dynamic [`Applier`](https://docs.rs/egg/latest/egg/trait.Applier.html) where you can have a right-hand side of rule that executes your own Rust code.

**Reply by big91987** [reply]
thx a lot! I'll try it
